Mutual Information estimator based on the Renyi quadratic entropy and th...

cdamon opened the pull request:
...e Cauchy Schwartz divergence combined with the Parzen window density estimator for continuous variable.

Work in progress for Parallelization Code with joblib and improve the function for MI of multiple features

Commit:
Mutual Information estimator based on the Renyi quadratic entropy and the Cauchy Schwartz divergence combined with the Parzen window density estimator for continuous variable.
Work in progress for Parallelization Code with joblib and improve the function for MI of multiple features

amueller commented:
Thanks for the pull request.
So if this is in the clustering module, it should have the interface of a clustering algorithm.
For that you'd need to provide a class that implements the fit function that computes the cluster assignments on some training data. See http://scikit-learn.org/dev/developers/index.html in particular http://scikit-learn.org/dev/developers/index.html#apis-of-scikit-learn-objects

Do you have a reference for using this MI estimate for clustering?

cdamon commented:
Sorry, I didn't know where to put the code and I did not receive an indication about it. It is probably not the right directory.
It is a metric useful for features selection to explore the relations between features and more specifically with the output  such as correlation metrics, Anova test, ...

amueller commented:
Well, as I said on the mailing list, do you have any good reference for the use as feature selection method, and its comparison to other methods? Please have a look at the interface of the classes in the feature_selection module.

cdamon commented:
I put some references in the code: 
For the mutual information estimation, the main reference is - D. XU and D. Erdogmuns, Renyi's entropy, divergence and their nonparametric estimators
(J.C. Principe, Information Theoretic Learning: Renyi’s Entropy and Kernel 47
Perspectives, Information Science and Statistics, DOI 10.1007/978-1-4419-1570-2 2, Springer Science+Business Media, LLC 2010)
For Features selection based on this mutual information, below are some references :
- Peng et al: feature selection based on mutual information : criteria of max-dependency, max-relevance and min-redundancy (IEEE transactions on pattern analysis and machine intelligence, 2005)
- Kari Torkkola, Feature Extraction by Non-Parametric Mutual Information
Maximization (Journal of Machine Learning Research 3 (2003) 1415-1438, http://www.jmlr.org/papers/volume3/torkkola03a/torkkola03a.pdf)
- Leonardo Macrini, Leonardo Gonçalves, Application of Rényi Entropy and Mutual Information
of Cauchy-Schwartz in Selecting Variables

It is also possible to use standard greedy algorithm for fetaures selection based on MI. 
I will look at the interface of the classes in the feature_selection module and try to compare a feature selection method based on MI with other methods. Do you think about a specific comparison?

Mutual information estimation may be integrated with other metrics of the same kind since it may be used for different purposes.

cdamon commented:
Moreover, I made some comparisons with other MI estimation obtained with histogram, Kraskov or knn aproaches and the proposed estimation based on Renyi entropy and Cauchy Schwartz divergence seems to be more accurate.

Commit:
a) Move MI_RenyiCC from  sklearn/metrics/cluster/ to sklearn/feature_selection/.

b) MI_RenyiCC.py includes MI estimation and functions for features selection as those described in univariate_selection. Since the previous version, I developped new features :
    1) estimation in  multi-dimensional cases (and not only bivariate variable) with both continuous, discrete variables or only continuous or discrete variable
    2) MI Estimation of multi-dimensional variable with respect to a discrete traget for classification purposes
    3) Code Parallelization with joblib
    4) Features selection functions  : univariate, multivariate, forward and backward algorithms.

c) test_MIRenyiCC.py describes examples to test features_selection based on MI estimation.

cdamon commented:
a) Move MI_RenyiCC from sklearn/metrics/cluster/ to sklearn/feature_selection/ as advised.  

b) MI_RenyiCC.py includes MI estimation and functions for features selection as those described in univariate_selection. Since the previous version, I developped new features : 

1) estimation in multi-dimensional cases (and not only bivariate variable) with both continuous, discrete variables or only continuous or discrete variable 
2) MI Estimation of multi-dimensional variable with respect to a discrete traget for classification purposes 
3) Code Parallelization with joblib 
4) You can estimate MI measure based on the k nearest neighbours to accelerate the computation. It works fine according to the number of neighbours considered (For instance, for iris data, if we had 16 noisy variables to the initial iris dataset of 150 samples, Univariate features selection based on MI by considering the 30 nearest neighbours allows to select the good features).
5) Features selection functions : univariate, multivariate, forward and backward algorithms.  

c) test_MIRenyiCC.py describes examples to test features_selection based on MI estimation. 

I will continue to add some test examples. 
Tell me if it seems more interesting for you and if you note errors or if you can improve it. It is really a big work and more and more excellent papers describe this MI estimation.

agramfort commented:
@cdamon please have a look at coding rules of sklearn and run flake8 on your code.

there are many style issues even before reviewing any algorithmic code

Commit:
AutoFlakes Run corrections

cdamon commented:
I understand. I have yet looked at coding rules (specially to optimize and parallelize the code and use scikit-learn functions and usages) but I probably miss some of them ;-). I will take some time to review them step by step.

Commit:
Using the threading backend of joblib to go faster

cdamon commented:
Hello, I added an important commit accelerating greatly the computation thanks to the threading backend of joblib. I do not understand the previous message "This branch has pending checks, but can be merged." I am not used with git project. Should I merge the branch? 
(git merge origin/ MI_RenyiCCParzen ?) 
Thanks

amueller commented:
Usually you should not merge a branch into a pull request. What exactly are you trying to do?

cdamon commented:
When I commited the last change yesterday I saw  The previous Travis CI build message which noticed that "This branch has pending checks, but can be merged." So I was wondering if I have to merge the branch. 
I have to look in more details at the coding rule as suggested by Alexandre Gramfort in its last message before you review the algorithm code if I have understood correctly. I will do it asap.
However, I don't know if you are interested by this work and if one is willing to contribute to its improvement.

amueller commented:
Unfortunately I can't answer the last question at the moment. There is a flurry of activity currently because Google Summer of Code application deadline is tomorrow, and we are generally short on reviewers at the moment.

amueller commented:
Also don't worry about what travis says, apart from the fact that it is failing ;)

cdamon commented:
Thanks for your reply. Ok I'll continue to review the code while awaiting your reviews or comments of the scikit-learn community. Concerning what Travis says, it is failing since I added the threading backend of joblib in the parallelization code. I don't know why.

lorentzenchr commented:
While some references are often cited, like Peng et al., the ones about Renyi entropy are not and therefore do not meet our [inclusion criteria](https://scikit-learn.org/stable/faq.html#what-are-the-inclusion-criteria-for-new-algorithms).
Also, this PR seems abandoned.
